types/react-redux/index.d.ts

```diff
@@ -19,16 +19,8 @@ export interface DispatchProp<S> {
   dispatch: Dispatch<S>;
 }
 
-interface ComponentDecorator<TMergedProps, TOwnProps> {
-    (component: Component<TOwnProps & TMergedProps>): ComponentClass<TOwnProps>;
-}
-
-interface ComponentDecoratorInfer<TMergedProps> {
-    <T>(component: Component<T & TMergedProps>): ComponentClass<T>;
-}
-
-interface ComponentMergeDecorator<TMergedProps, TOwnProps> {
-    (component: Component<TMergedProps>): ComponentClass<TOwnProps>;
+export interface WrapWithDispatch {
+    <P, TComponent extends Component<P>>(component: TComponent): TComponent;
```

Comment:
The returned component is typed the same as the component passed in to this method as if nothing was changed on the original component. Which is any object that extends `ComponentClass<P> | StatelessComponent<P>`.